Prevent unwanted helper api calls when loading subscription notes

thilinah opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Connected WooCommerce stores send requests to helper API for fetching details related to subscriptions and to get information about plugin updates. The responses from these API endpoints are cached to prevent stores from sending multiple requests frequently. 


For example, the `GET subscriptions` request is cached for  1 hour and `POST update-check` request is cached for 12 hours.

But for some stores, the cache can be completely broken due to DB having a full disk. This is because when `set_transient` is called if the object cache is not configured for the site wp options table will be used to store the transient value. (further discussions in this [P2 post](https://wp.me/p6JqRr-4N1#comment-12553) and the [issue](https://github.com/Automattic/woocommerce.com/issues/11649)).

#### Problem

- Subscription notes are loaded via `admin_init`
- In order to prevent subscription notes from on every call to `admin_init` the option `LAST_REFRESH_OPTION_KEY` is used.
- But when the db disk is full, saving the latest updated time in wp_option should fail.
- This leads to subscription notes being loaded on every admin page load and cron run.
- Since the transients are not saved when the disk is full, this leads to store making get subscription API call multiple times.

Fix
- There is no need to check if subscription notes need to be refreshed on every admin page load. Instead, it can be checked only when the user visits a WC page showing the `Activity` tab.


Closes # .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

#### Cron should not load subscription notes
- Make sure the site is connected to WCCOM via <your test site root>wp-admin/admin.php?page=wc-addons&section=helper
- Every time subscription notes are updated the option `woocommerce_admin-wc-helper-last-refresh` will be set to the current time. We will use this fact for testing.
- Checkout trunk.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Run cron `wp cron event run --all`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it exists.
- Checkout this branch.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Run cron `wp cron event run --all`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it doesn't exist.

#### Only visiting a WC Admin or a page with WC Admin Header Shall load subscription notes
- Checkout this branch.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Visit a non WC page like `wp-admin/edit-comments.php`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it doesn't exist.
- Visit a WC related page that shows `Activity` tab, such as `wp-admin/edit.php?post_type=shop_order`
![Screenshot 2023-06-02 at 12 47 59](https://github.com/woocommerce/woocommerce/assets/1531403/62dfce78-c733-424f-93b6-3212a3e2b6f1)
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice that it exists. (this means subscription notes are loading on this page)
- Repeat this test on any page to make sure all pages showing `Activity` tab are loading subscription notes.

#### Make sure the subscription notes are loading after this update
- Checkout this branch.
- Make sure your store is connected.
- You need to have an active product that is expiring soon (less than in 60 days)
- If this is not the case, you may update the expiry date of the subscription via the dashboard.
- Under [My Subscriptions](https://wp.test:4443/wp-admin/admin.php?page=wc-addons&section=helper) you should be able to see the expiry date.
![Screenshot 2023-06-02 at 12 41 41](https://github.com/woocommerce/woocommerce/assets/1531403/72f437db-a72f-4b6d-8bdd-5f599eded92f)
- Disable the auto-renew for this product via [My Subscriptions](https://woocommerce.test/my-account/my-subscriptions/)
- Run the following in `wp shell` to remove subscription notes.
```
\Automattic\WooCommerce\Admin\Notes\Notes::delete_notes_with_name('wc-admin-wc-helper-subscription');
```
- Update subscriptions under [My Subscriptions](https://wp.test:4443/wp-admin/admin.php?page=wc-addons&section=helper) 
![Screenshot 2023-06-02 at 12 41 55](https://github.com/woocommerce/woocommerce/assets/1531403/c92c0379-8c84-4fc3-bda3-853d18821d24)
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh` reset last subscription notes refresh time.
- Load any WC page.
- Notice there is a new subscription note as a reminder to enable auto-renew for the above mentioned product.
![Screenshot 2023-06-02 at 12 41 21](https://github.com/woocommerce/woocommerce/assets/1531403/78982bb4-6308-4ae5-b8f2-1cc12f6797a5)


<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

thilinah committed:
Load subscription notes only via WC Admin home page and extensions page to avoid calling helper API endpoints unnecessarily.

thilinah committed:
Adding change log.

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php:
```diff
@@ -80,6 +80,17 @@ public function update_option_woocommerce_helper_data( $old_value, $value ) {
 	public function admin_init() {
 		$this->check_connection();
```
seems like this isn't required either, perhaps we can move this below the early return statements?

vedanshujain reviewed (dismissed):
Requesting a minor change if its possible

thilinah committed:
Moving connection check after the early return statements.

thilinah committed:
Merge branch 'trunk' into fix/unwanted-helper-api-calls-when-loading-subscription-notes

vedanshujain commented:
@thilinah This looks good to me, but I am assuming you are waiting for wc-admin folks to also take a look?

rcstr reviewed (dismissed):
Changes looks good and work as expected

thilinah commented:
> @thilinah This looks good to me, but I am assuming you are waiting for wc-admin folks to also take a look?

@vedanshujain if the changes look good can you please approve the PR, I see it's blocked due to the change request.

vedanshujain commented:
@thilinah dismissed my review, should be mergeable now

denho commented:
@thilinah The WC notes are shown not only on `wc-admin` and `wc-addons` pages. They are also loaded on other WC pages in admin, such as orders list or settings.  
<img width="386" alt="image" src="https://github.com/woocommerce/woocommerce/assets/2803994/c1e5d142-fc47-41cc-bd16-8c52e3f18ad0">
On those pages, the subscription notes are not refreshed due to changes in your PR, and the user will not get the latest  notes until they visit `wc-admin` and `wc-addons` pages.

thilinah commented:
> On those pages, the subscription notes are not refreshed due to changes in your PR, and the user will not get the latest notes until they visit wc-admin and wc-addons pages.

@denho in this case I think skipping subscription notes loading based on the page won't work. I've noticed that other WC plugins like subscriptions and bookings also show the `Activity` button. So post_types such as `shop_order`, `shop_subscription`, `shop_coupon`, `product`. 

Due to this, I think we should consider moving forward with https://github.com/woocommerce/woocommerce/pull/36846 and move the discussion to that PR for refining the approach.

thilinah commented:
Also, another way could be to make sure subscription notes are not loaded via the cron.

thilinah committed:
Use admin_head hook instead of admin_init to load subscriptions notes.

Since `get_current_screen` function can not be called within `admin-init` hook it is not possible to determine if the subscriptions are being loaded via WC related page or not.

thilinah commented:
@denho @nirolph @rcstr re-requesting reviews for this PR since the code and test instructions are updated.

rcstr reviewed (approved):
Changes looks good and works as described

jonathansadowski commented:
I'm going to clear the milestone from this for now, so that this doesn't inadvertently get missed once it's merged. A milestone will be added automatically by our automation upon merge.

denho commented on plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php:
```diff
@@ -75,9 +75,16 @@ public function update_option_woocommerce_helper_data( $old_value, $value ) {
 	 */
-	public function admin_init() {
+	public function admin_head() {
+		if ( ! \Automattic\WooCommerce\Admin\PageController::is_admin_or_embed_page() ) {
```
How about adding `use \Automattic\WooCommerce\Admin\PageController` and replacing this line with 
```suggestion
		if ( ! PageController::is_admin_or_embed_page() ) {
```

denho commented on plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php:
```diff
@@ -75,9 +75,16 @@ public function update_option_woocommerce_helper_data( $old_value, $value ) {
 	/**
-	 * Things to do on admin_init.
+	 * Things to do on admin_head.
 	 */
-	public function admin_init() {
+	public function admin_head() {
```
Would be good to have a more meaningful description of the method functionality in the DocBlock, and method name.

denho reviewed (approved):
Added some suggestions in the code. Otherwise, LGTM!

thilinah committed:
Update doc block.

thilinah commented:
@jonathansadowski, the PR is approved. Is it possible to get this merged?

jonathansadowski commented:
Hi @thilinah, please feel free to merge once checks have passed (as they have in this case), and you've gotten your PR approved. If you don't have access to merge yourself, let me know and I can take a look further.

thilinah commented:
@jonathansadowski I'm not authorized to merge.

jonathansadowski commented:
Thanks @thilinah, I've merged this for now — I'll look into whether or not that should be the case.
